File: tests/Unit/Evolution/Systems/ScalarWave/Test_SphericalShellPowerMonitor.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Framework/TestingFramework.hpp"

#include <array>
#include <cmath>
#include <cstddef>

#include "DataStructures/DataVector.hpp"
#include "DataStructures/Index.hpp"
#include "DataStructures/ModalVector.hpp"
#include "DataStructures/Tensor/TypeAliases.hpp"
#include "DataStructures/Variables.hpp"
#include "Evolution/Systems/ScalarWave/ApplyTensorYlmFilter.hpp"
#include "Evolution/Systems/ScalarWave/SphericalShellPowerMonitor.hpp"
#include "Evolution/Systems/ScalarWave/Tags.hpp"
#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
#include "NumericalAlgorithms/Spectral/Mesh.hpp"
#include "Utilities/Gsl.hpp"

namespace {

Mesh<3> shell_mesh(const size_t radial_extents, const size_t ell_max) {
  return Mesh<3>{
      {{radial_extents, ell_max + 1, 2 * ell_max + 1}},
      {{Spectral::Basis::Legendre, Spectral::Basis::SphericalHarmonic,
        Spectral::Basis::SphericalHarmonic}},
      {{Spectral::Quadrature::GaussLobatto, Spectral::Quadrature::Gauss,
        Spectral::Quadrature::Equiangular}}};
}

InverseJacobian<DataVector, 3, Frame::Inertial, Frame::Grid> identity_jacobian(
    const size_t size) {
  InverseJacobian<DataVector, 3, Frame::Inertial, Frame::Grid> jacobian{size,
                                                                        0.0};
  for (size_t i = 0; i < 3; ++i) {
    jacobian.get(i, i) = 1.0;
  }
  return jacobian;
}

void set_constant_angular_radial_profile(
    const gsl::not_null<DataVector*> component, const Mesh<3>& mesh,
    const DataVector& radial_profile) {
  const auto extents = mesh.extents();
  for (size_t r = 0; r < extents[0]; ++r) {
    for (size_t theta = 0; theta < extents[1]; ++theta) {
      for (size_t phi = 0; phi < extents[2]; ++phi) {
        Index<3> index{};
        index[0] = r;
        index[1] = theta;
        index[2] = phi;
        (*component)[collapsed_index(index, extents)] = radial_profile[r];
      }
    }
  }
}

// The radial monitor for a single tensor component set to a radial profile
// (constant in angle) is |a_r| / sqrt(num_tensor_components), where a_r are
// the radial modal coefficients.
DataVector expected_radial_monitor(const Mesh<3>& mesh,
                                   const DataVector& radial_profile,
                                   const size_t num_tensor_components) {
  ModalVector radial_modes(radial_profile.size(), 0.0);
  to_modal_coefficients(make_not_null(&radial_modes), radial_profile,
                        mesh.slice_through(0));
  DataVector expected(radial_profile.size(), 0.0);
  for (size_t i = 0; i < expected.size(); ++i) {
    expected[i] = std::abs(radial_modes[i]) /
                  sqrt(static_cast<double>(num_tensor_components));
  }
  return expected;
}

}  // namespace

SPECTRE_TEST_CASE(
    "Unit.Evolution.Systems.ScalarWave.SphericalShellPowerMonitor",
    "[Unit][Evolution]") {
  const Mesh<3> mesh = shell_mesh(4, 3);
  Variables<ylm::TensorYlm::filter_detail::sw_vars_list<Frame::Inertial>>
      sw_vars(mesh.number_of_grid_points(), 0.0);
  const DataVector radial_profile{1.0, -0.5, 0.25, 2.0};
  ScalarWave::power_monitor::SwCartToSphereMatrix cart_to_sphere{};

  // Psi: 1 component, spin weight 0.  A constant-angular profile excites
  // only the l=0 SH mode; the radial monitor should match the modal
  // coefficients of the radial profile divided by sqrt(1).
  set_constant_angular_radial_profile(
      make_not_null(&get(get<ScalarWave::Tags::Psi>(sw_vars))), mesh,
      radial_profile);
  {
    const auto monitors = ScalarWave::power_monitor::sw_shell_power_monitors(
        make_not_null(&cart_to_sphere), sw_vars, mesh,
        identity_jacobian(mesh.number_of_grid_points()));
    CHECK_ITERABLE_APPROX(monitors.psi[0],
                          expected_radial_monitor(mesh, radial_profile, 1));
    CHECK(max(monitors.psi[1]) > 0.0);
    CHECK(max(abs(monitors.pi[0])) == 0.0);
    CHECK(max(abs(monitors.pi[1])) == 0.0);
    CHECK(max(abs(monitors.phi[0])) == 0.0);
    CHECK(max(abs(monitors.phi[1])) == 0.0);
  }

  // Pi: 1 component, spin weight 0.  Same structure as Psi.
  std::fill(sw_vars.data(), sw_vars.data() + sw_vars.size(), 0.0);
  set_constant_angular_radial_profile(
      make_not_null(&get(get<ScalarWave::Tags::Pi>(sw_vars))), mesh,
      radial_profile);
  {
    const auto monitors = ScalarWave::power_monitor::sw_shell_power_monitors(
        make_not_null(&cart_to_sphere), sw_vars, mesh,
        identity_jacobian(mesh.number_of_grid_points()));
    CHECK_ITERABLE_APPROX(monitors.pi[0],
                          expected_radial_monitor(mesh, radial_profile, 1));
    CHECK(max(monitors.pi[1]) > 0.0);
    CHECK(max(abs(monitors.psi[0])) == 0.0);
    CHECK(max(abs(monitors.psi[1])) == 0.0);
    CHECK(max(abs(monitors.phi[0])) == 0.0);
    CHECK(max(abs(monitors.phi[1])) == 0.0);
  }

  // Phi: 3 components (tnsr::i<DataVector,3>), spin weights 0 and +/-1.
  // Set Phi_x only; the cart-to-sphere transform distributes power across
  // the spin-weighted components, so only the radial monitor (which is
  // computed from the raw un-transformed data) has the simple form
  // |a_r| / sqrt(3).  Psi and Pi must remain zero.
  std::fill(sw_vars.data(), sw_vars.data() + sw_vars.size(), 0.0);
  set_constant_angular_radial_profile(
      make_not_null(&get<0>(get<ScalarWave::Tags::Phi<3>>(sw_vars))), mesh,
      radial_profile);
  {
    const auto monitors = ScalarWave::power_monitor::sw_shell_power_monitors(
        make_not_null(&cart_to_sphere), sw_vars, mesh,
        identity_jacobian(mesh.number_of_grid_points()));
    CHECK_ITERABLE_APPROX(monitors.phi[0],
                          expected_radial_monitor(mesh, radial_profile, 3));
    CHECK(max(monitors.phi[1]) > 0.0);
    CHECK(max(abs(monitors.psi[0])) == 0.0);
    CHECK(max(abs(monitors.psi[1])) == 0.0);
    CHECK(max(abs(monitors.pi[0])) == 0.0);
    CHECK(max(abs(monitors.pi[1])) == 0.0);
  }
}

File: tests/Unit/Evolution/Systems/ScalarWave/Python/Test_SphericalShellPowerMonitor.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import unittest

import numpy as np
import numpy.testing as npt

from spectre.DataStructures import DataVector
from spectre.DataStructures.Tensor import Frame, Scalar, tnsr
from spectre.Domain import ElementId
from spectre.Domain.Creators import SphericalShells
from spectre.Evolution.Systems.ScalarWave import (
    sw_b3_power_monitors,
    sw_shell_power_monitors,
)
from spectre.Spectral import Basis, Mesh, Quadrature, logical_coordinates


class TestSwSphericalShellPowerMonitor(unittest.TestCase):
    def setUp(self):
        self.inner_radius = 4.0
        self.outer_radius = 7.0
        self.radial_points = 5
        self.l_max = 3
        self.mesh = Mesh[3](
            [self.radial_points, self.l_max + 1, 2 * self.l_max + 1],
            [
                Basis.Legendre,
                Basis.SphericalHarmonic,
                Basis.SphericalHarmonic,
            ],
            [
                Quadrature.GaussLobatto,
                Quadrature.Gauss,
                Quadrature.Equiangular,
            ],
        )
        self.domain = SphericalShells(
            self.inner_radius,
            self.outer_radius,
            0,
            self.radial_points,
            self.l_max,
        ).create_domain()
        self.element_id = ElementId[3](0)
        self.num_points = self.mesh.number_of_grid_points()

    def _zero_fields(self):
        psi = Scalar[DataVector](num_points=self.num_points, fill=0.0)
        pi = Scalar[DataVector](num_points=self.num_points, fill=0.0)
        phi = tnsr.i[DataVector, 3, Frame.Inertial](
            num_points=self.num_points, fill=0.0
        )
        return psi, pi, phi

    def _call(self, psi, pi, phi):
        return sw_shell_power_monitors(
            psi,
            pi,
            phi,
            self.mesh,
            self.element_id,
            self.domain,
            0.0,
            {},
        )

    def test_result_structure(self):
        """Result dict has the expected keys and array lengths."""
        psi, pi, phi = self._zero_fields()
        monitors = self._call(psi, pi, phi)

        self.assertEqual(set(monitors.keys()), {"Psi", "Pi", "Phi"})
        for variable in ("Psi", "Pi", "Phi"):
            self.assertEqual(
                set(monitors[variable].keys()), {"radial", "angular"}
            )
            self.assertEqual(
                len(monitors[variable]["radial"]), self.radial_points
            )
            self.assertEqual(len(monitors[variable]["angular"]), self.l_max + 1)

    def test_psi_isolation(self):
        """Setting only Psi gives zero Pi and Phi monitors."""
        psi, pi, phi = self._zero_fields()
        logical_coords = np.asarray(logical_coordinates(self.mesh))
        # Purely radial profile: uniform in angle (excites only l=0 SH mode).
        radial = DataVector(
            0.5 * (self.outer_radius - self.inner_radius) * logical_coords[0]
            + 0.5 * (self.outer_radius + self.inner_radius)
        )
        psi[0] = radial

        monitors = self._call(psi, pi, phi)

        self.assertGreater(max(monitors["Psi"]["radial"]), 0.0)
        self.assertGreater(monitors["Psi"]["angular"][0], 0.0)
        npt.assert_array_equal(np.asarray(monitors["Pi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Pi"]["angular"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Phi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Phi"]["angular"]), 0.0)

    def test_pi_isolation(self):
        """Setting only Pi gives zero Psi and Phi monitors."""
        psi, pi, phi = self._zero_fields()
        logical_coords = np.asarray(logical_coordinates(self.mesh))
        radial = DataVector(
            0.5 * (self.outer_radius - self.inner_radius) * logical_coords[0]
            + 0.5 * (self.outer_radius + self.inner_radius)
        )
        pi[0] = radial

        monitors = self._call(psi, pi, phi)

        self.assertGreater(max(monitors["Pi"]["radial"]), 0.0)
        self.assertGreater(monitors["Pi"]["angular"][0], 0.0)
        npt.assert_array_equal(np.asarray(monitors["Psi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Psi"]["angular"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Phi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Phi"]["angular"]), 0.0)

    def test_phi_isolation(self):
        """Setting only Phi_x gives zero Psi and Pi monitors."""
        psi, pi, phi = self._zero_fields()
        logical_coords = np.asarray(logical_coordinates(self.mesh))
        radial = DataVector(
            0.5 * (self.outer_radius - self.inner_radius) * logical_coords[0]
            + 0.5 * (self.outer_radius + self.inner_radius)
        )
        phi[phi.get_storage_index(0)] = radial

        monitors = self._call(psi, pi, phi)

        self.assertGreater(max(monitors["Phi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Psi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Psi"]["angular"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Pi"]["radial"]), 0.0)
        npt.assert_array_equal(np.asarray(monitors["Pi"]["angular"]), 0.0)

    def test_psi_pi_equal_for_same_profile(self):
        """Psi and Pi monitors are identical when given the same profile."""
        logical_coords = np.asarray(logical_coordinates(self.mesh))
        profile = DataVector(
            0.5 * (self.outer_radius - self.inner_radius) * logical_coords[0]
            + 0.5 * (self.outer_radius + self.inner_radius)
        )

        psi, pi, phi = self._zero_fields()
        psi[0] = profile
        m_psi = self._call(psi, pi, phi)

        psi2, pi2, phi2 = self._zero_fields()
        pi2[0] = profile
        m_pi = self._call(psi2, pi2, phi2)

        npt.assert_allclose(
            np.asarray(m_psi["Psi"]["radial"]),
            np.asarray(m_pi["Pi"]["radial"]),
            rtol=1.0e-13,
        )
        npt.assert_allclose(
            np.asarray(m_psi["Psi"]["angular"]),
            np.asarray(m_pi["Pi"]["angular"]),
            rtol=1.0e-13,
        )

    def test_incompatible_mesh_raises(self):
        """A Legendre-only mesh raises RuntimeError."""
        psi, pi, phi = self._zero_fields()
        bad_mesh = Mesh[3](
            self.mesh.extents(),
            [Basis.Legendre, Basis.Legendre, Basis.Legendre],
            [
                Quadrature.GaussLobatto,
                Quadrature.GaussLobatto,
                Quadrature.GaussLobatto,
            ],
        )
        with self.assertRaisesRegex(
            RuntimeError, "require the mesh dimensions"
        ):
            sw_shell_power_monitors(
                psi,
                pi,
                phi,
                bad_mesh,
                self.element_id,
                self.domain,
                0.0,
                {},
            )

    def test_b3_result_structure(self):
        """sw_b3_power_monitors returns the expected dict structure."""
        n_r = 3
        l_max = 2
        b3_mesh = Mesh[3](
            [n_r, l_max + 1, 2 * l_max + 1],
            [Basis.ZernikeB3, Basis.ZernikeB3, Basis.ZernikeB3],
            [
                Quadrature.GaussRadauUpper,
                Quadrature.Gauss,
                Quadrature.Equiangular,
            ],
        )
        num_points = b3_mesh.number_of_grid_points()
        psi = Scalar[DataVector](num_points=num_points, fill=1.0)
        pi = Scalar[DataVector](num_points=num_points, fill=0.0)
        phi = tnsr.i[DataVector, 3, Frame.Inertial](
            num_points=num_points, fill=0.0
        )

        monitors = sw_b3_power_monitors(
            psi,
            pi,
            phi,
            b3_mesh,
            self.element_id,
            self.domain,
            0.0,
            {},
        )

        self.assertEqual(set(monitors.keys()), {"Psi", "Pi", "Phi"})
        for variable in ("Psi", "Pi", "Phi"):
            self.assertEqual(
                set(monitors[variable].keys()), {"radial", "angular"}
            )
            self.assertEqual(len(monitors[variable]["radial"]), n_r)
            self.assertEqual(len(monitors[variable]["angular"]), l_max + 1)


if __name__ == "__main__":
    unittest.main(verbosity=2)
